inspectkv/inspectkv.go

```diff
@@ -158,7 +288,89 @@ func ScanSnapshotTableData(store kv.Storage, ver kv.Version, t table.Table, star
 	}
 	defer snap.Release()
 
-	records, nextHandle, err := ScanTableData(t, snap, startHandle, limit)
+	records, nextHandle, err := ScanTableData(snap, t, startHandle, limit)
 
 	return records, nextHandle, errors.Trace(err)
 }
+
+// CompareTableData compares data and the corresponding table data one by one.
+// It returns nil if data is equal to the data that scans from table, otherwise
+// it returns an error with a different set of records. If exact is false, only
+// compares handle.
+func CompareTableData(txn kv.Transaction, t table.Table, data []*RecordData, exact bool) error {
+	var err error
+	var vals []interface{}
+	m := make(map[int64]struct{}, len(data))
+
+	for _, r := range data {
+		vals, err = t.RowWithCols(txn, r.Handle, t.Cols())
+		if terror.ErrorEqual(err, kv.ErrNotExist) {
+			record1 := &RecordData{Handle: r.Handle, Values: r.Values}
+			err = errors.Errorf("data:%v != record:%v", record1, nil)
+		}
+		if err != nil {
+			break
+		}
+
+		if !exact {
+			m[r.Handle] = struct{}{}
+			continue
+		}
+		if !reflect.DeepEqual(r.Values, vals) {
+			record1 := &RecordData{Handle: r.Handle, Values: r.Values}
+			record2 := &RecordData{Handle: r.Handle, Values: vals}
+			err = errors.Errorf("data:%v != record:%v", record1, record2)
+			break
+		}
+		m[r.Handle] = struct{}{}
+	}
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	startKey := t.RecordKey(0, nil)
+	filterFunc := func(h int64, vals []interface{}, cols []*column.Col) (bool, error) {
+		if _, ok := m[h]; !ok {
+			record := &RecordData{Handle: h, Values: vals}
+			return false, errors.Errorf("data:%v != record:%v", nil, record)
+
+		}
+
+		return true, nil
+	}
+	err = t.IterRecords(txn, string(startKey), t.Cols(), filterFunc)
```

Comment:
Why not checking result only in IterRecords loop, so that u will not call RowWithCols so many times.